File: 0219-contains-duplicate-ii/0219-contains-duplicate-ii.java
class Solution {
    public boolean containsNearbyDuplicate(int[] nums, int k) {
        HashSet<Integer> hset = new HashSet<Integer>();
        final int n = nums.length;
        //first window
        for(int i = 0; i < Math.min(k + 1, n); i++) {
            if(hset.contains(nums[i])) {
                return true;
            }
            hset.add(nums[i]);
        }
        for(int i = k + 1; i < n; i++) {
            int j = i - k - 1;
            hset.remove(nums[j]);
            if(hset.contains(nums[i])) {
                return true;
            }
            hset.add(nums[i]);
        }
        return false;
    }
}

File: 0219-contains-duplicate-ii/0219-contains-duplicate-ii.js
/**
 * @param {number[]} nums
 * @param {number} k
 * @return {boolean}
 */
var containsNearbyDuplicate = function(nums, k) {
    
    const n = nums.length;
    let set = new Set();
    for(let i = 0; i < n; i++) {
        if(i > k) {
            set.delete(nums[i - k - 1]);
        }
        if(set.has(nums[i])) return true
        else set.add(nums[i]);
    }
    return false;
};